src/coreclr/src/jit/flowgraph.cpp

```diff
@@ -3557,10 +3557,165 @@ void Compiler::fgInitBlockVarSets()
     fgBBVarSetsInited = true;
 }
 
+//------------------------------------------------------------------------------
+// fgInsertGCPolls : Insert GC polls for basic blocks containing calls to methods
+//                   with SuppressGCTransitionAttribute.
+//
+//
+// Notes:
+//    When not optimizing, the method relies on BBF_HAS_SUPPRESSGC_CALL flag to
+//    find the basic blocks that require GC pols; when optimizing the tree nodes
+//    are scanned to find calls to methods with SuppressGCTransitionAttribute.
+//
+
+void Compiler::fgInsertGCPolls()
+{
+    if ((optMethodFlags & OMF_NEEDS_GCPOLLS) == 0)
+    {
+        return;
+    }
+
+    bool createdPollBlocks = false;
+
+#ifdef DEBUG
+    if (verbose)
+    {
+        printf("*************** In fgInsertGCPolls() for %s\n", info.compFullName);
+        fgDispBasicBlocks(false);
+        printf("\n");
+    }
+#endif // DEBUG
+
+    BasicBlock* block;
+
+    // Walk through the blocks and hunt for a block that has needs a GC Poll
+    for (block = fgFirstBB; block; block = block->bbNext)
+    {
+        bool blockNeedsGCPoll = false;
+        if (opts.OptimizationDisabled())
+        {
+            if ((block->bbFlags & BBF_HAS_SUPPRESSGC_CALL) != 0)
+            {
+                blockNeedsGCPoll = true;
+            }
+        }
+        else
+        {
+            // When optimizations are enabled, we can't rely on BBF_HAS_SUPPRESSGC_CALL flag:
+            // the call could've been moved, e.g., hoisted from a loop, CSE'd, etc.
+            for (Statement* stmt = block->FirstNonPhiDef(); !blockNeedsGCPoll && (stmt != nullptr);
+                 stmt            = stmt->GetNextStmt())
+            {
+                for (GenTree* tree = stmt->GetTreeList(); !blockNeedsGCPoll && (tree != nullptr); tree = tree->gtNext)
```

Comment:
Can we trust top-level `gtFlags` to have a properly up to date `GTF_CALL` and bypass some of this searching?

Reply:
Great suggestion, done. Yes, we can trust top-level `gtFlags` to have a properly up to date `GTF_CALL`: we check that in `fgDebugCheckFlags`.